### ISSUE TITLE
How many of the bytes served out of s3 go to bots? How much does that cost?

### ISSUE THREAD

**Issue by eddierubeiz** [sciencehistory/scihist_digicoll#1615]
TL;DR:

- On a sample day picked at random, 78% of the information served from s3 went to bots.
- On that day, the bot traffic cost us on the order of $10.
- We now have recipes we can use to dig deeper if we need.

----

Last week we figured out how to turn on rudimentary logging for s3 (see https://github.com/sciencehistory/terraform_scihist_digicoll/issues/8)

I'm creating this ticket to store some basic research that can be used to inform later decisions on how we manage s3 traffic in the digital collections. For now, we're focusing on our largest and most expensive bucket, the production originals bucket, and using the user agent string to guess whether a given request was served to a bot.

**Comment by eddierubeiz** [comment]
For future reference, here's the `bash` code I used to extract the data for a sample day (last Saturday)
```bash
DATE='*2022-03-19*'
BUCKET="s3_server_access_production_originals"
aws s3 sync s3://chf-logs/s3_server_access_production_originals        s3_server_access_production_originals
for FILE in `find s3_server_access_production_originals -name "$DATE"`; do	
   cat $FILE | while read LINE; do
      BYTES_SENT=`echo $LINE | cut -d ' ' -f 15`
      if [ "$BYTES_SENT" != '-' ]; then
         USER_AGENT=`echo $LINE | cut -d '"' -f 6`
         echo "\"$BYTES_SENT\",\"$USER_AGENT\""
      fi
   done
done
```

**Comment by eddierubeiz** [comment]
The results of that script can be fed into a pivot table and totaled up by user agent. Our main bots appear to be Applebot, Googlebot, SemrushBot, Gootlebot-Image.

**Comment by eddierubeiz** [comment]
The total bytes served is `176068528846` bytes according to the recipe above: and `183926581202` according to the s3 dashboard.
We assume the 4% discrepancy is fine since
 - s3 logging is ["best effort"](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerLogs.html), and 
 - our recipe is filtering out the day based on the date in the filename, not the log lines in the files.
For our sample day, our traffic is roughly 78% bots, 22% everything else.

**Comment by CHF-IT-Staff** [comment]
There could also be a time zone issue causing that 4%.  AWS is using UTC for its bytes calculation


__________________________________

Chuck Hornstein   (he, him, his)
Director, Information Technology
t. +1.215.873.8272
f. +1.215.629.5272

Science History Institute
Chemistry * Engineering * Life Sciences
315 Chestnut Street * Philadelphia, PA 19106 * U.S.A.
Learn about the scientific discoveries that changed our world at sciencehistory.org/learn<https://www.sciencehistory.org/learn>

**Comment by eddierubeiz** [comment]
Good call.

**Comment by eddierubeiz** [comment]
No, the logs are explicitly in UTC as well: e.g. `17/Mar/2022:22:08:00 +0000` .

**Comment by CHF-IT-Staff** [comment]
Thanks Eddie


__________________________________

Chuck Hornstein   (he, him, his)
Director, Information Technology
t. +1.215.873.8272
f. +1.215.629.5272

Science History Institute
Chemistry * Engineering * Life Sciences
315 Chestnut Street * Philadelphia, PA 19106 * U.S.A.
Learn about the scientific discoveries that changed our world at sciencehistory.org/learn<https://www.sciencehistory.org/learn>

**Comment by eddierubeiz** [comment]
Amended the code to be slower and more precise. Edited the code above accordingly.
- New total of bytes served: `183828209511` bytes.
- Compare with `183926581202` bytes as reported by s3.
- Does not change the proportion of bot traffic substantially -- still roughly 78% of our prod original files are served to bots.

**Comment by eddierubeiz** [comment]
Note: at a cost of 9 cents per GB out, the cost for bytes served from that bucket should come out to $16.54 for last Saturday.

**Comment by eddierubeiz** [comment]
**Note:** the actual cost for our sample day is actually cheaper, at $14.70, but we know the 9¢/GB cost does not apply for all our outbound traffic:
- Requests from our own Heroku servers in `us-east-1`, such as our fixity checks, are free;
- Cross-region replication requests, while not free, are cheaper than 9¢/GB.
A more comprehensive model would take the IAM user (which is included in the logs) into account and could presumably get us much closer to the actual cost. That's out of scope for this ticket, however.

We know at this point that bots need to be told not to crawl our non-PDF originals.

**Comment by eddierubeiz** [comment]
You may be right. In practice can probably distinguish the signature of a request we **know** was a fixity check from one we **know** was a call to `/download` ...  if we decide it's important.
